File: src/exporter.ts
import type { GeminiMessage } from "./types/gemini.js";

/**
 * Options for chat export formatting.
 */
export interface ExportOptions {
  /** Include metadata (title, export date, message count) at the top of the export */
  includeMetadata?: boolean;
}

function formatContent(content: string): string {
  if (content.includes("```")) {
    return formatCodeBlocks(content);
  }
  return content;
}

function formatCodeBlocks(content: string): string {
  const lines = content.split("\n");
  const result: string[] = [];
  let inCodeBlock = false;
  let codeLanguage = "";

  for (const line of lines) {
    if (line.trim().startsWith("```")) {
      if (!inCodeBlock) {
        inCodeBlock = true;
        codeLanguage = line.trim().slice(3).trim();
        if (codeLanguage) {
          result.push(`\`\`\`${codeLanguage}`);
        } else {
          result.push("```");
        }
      } else {
        inCodeBlock = false;
        result.push("```");
      }
    } else if (inCodeBlock) {
      result.push(line);
    } else {
      result.push(line);
    }
  }

  return result.join("\n");
}

/**
 * Formats a list of Gemini messages as a Markdown document.
 * @param messages - Array of GeminiMessage objects to format
 * @param title - Title for the exported chat
 * @param conversationId - Optional conversation ID to include as metadata
 * @param options - Export options such as including metadata
 * @returns Formatted Markdown string
 */
export function formatChatAsMarkdown(
  messages: GeminiMessage[],
  title: string,
  conversationId?: string,
  options: ExportOptions = {}
): string {
  const exportDate = new Date().toISOString().replace("T", " ").slice(0, 19);
  const messageCount = messages.length;

  const lines: string[] = [];

  if (conversationId) {
    lines.push(`<!-- conversation_id: ${conversationId} -->`);
  }

  if (options.includeMetadata) {
    lines.push("---");
    lines.push(`title: "${title}"`);
    lines.push(`export_date: "${exportDate}"`);
    lines.push(`message_count: ${messageCount}`);
    lines.push("---\n");
  } else {
    lines.push("");
  }

  lines.push(`# ${title}\n`);

  for (const msg of messages) {
    const role = msg.role || "user";
    const content = msg.content || "";
    const timestamp = (msg as any).timestamp;

    if (role === "user") {
      lines.push("**User:**");
    } else {
      lines.push("**Gemini:**");
    }

    if (timestamp) {
      lines.push(`<small>${timestamp}</small>`);
    }

    lines.push("");
    lines.push(formatContent(content));
    lines.push("");
  }

  return lines.join("\n");
}

File: src/gemiterm/exporter.py
from datetime import datetime
from typing import Any


def format_chat_as_markdown(
    messages: list[dict[str, Any]],
    title: str,
    conversation_id: str | None = None,
    include_metadata: bool = False,
) -> str:
    export_date = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    message_count = len(messages)

    lines: list[str] = []

    if conversation_id:
        lines.append(f"<!-- conversation_id: {conversation_id} -->")

    if include_metadata:
        lines.append("---")
        lines.append(f'title: "{title}"')
        lines.append(f'export_date: "{export_date}"')
        lines.append(f"message_count: {message_count}")
        lines.append("---\n")
    else:
        lines.append("")

    lines.append(f"# {title}\n")

    for msg in messages:
        role = msg.get("role", "user")
        content = msg.get("content", "")
        timestamp = msg.get("timestamp")
        images = msg.get("images")
        citations = msg.get("citations")

        if role == "user":
            lines.append("**User:**")
        else:
            lines.append("**Gemini:**")

        if timestamp:
            lines.append(f"<small>{timestamp}</small>")

        lines.append("")
        lines.append(_format_content(content))
        lines.append("")

        if images:
            lines.append("")
            lines.append("**Images:**")
            for img in images:
                if isinstance(img, dict):
                    lines.append(f"- {img.get('title', 'Image')}: {img.get('url', '')}")
                else:
                    lines.append(f"- {img}")
            lines.append("")

        if citations:
            lines.append("")
            lines.append("**Citations:**")
            for cit in citations:
                if isinstance(cit, dict):
                    lines.append(f"- [{cit.get('title', 'Source')}]({cit.get('url', '')})")
                else:
                    lines.append(f"- {cit}")
            lines.append("")

    return "\n".join(lines)


def _format_content(content: str) -> str:
    if "```" in content:
        return _format_code_blocks(content)
    return content


def _format_code_blocks(content: str) -> str:
    lines = content.split("\n")
    result: list[str] = []
    in_code_block = False
    code_language = ""

    for line in lines:
        if line.strip().startswith("```"):
            if not in_code_block:
                in_code_block = True
                code_language = line.strip()[3:].strip()
                if code_language:
                    result.append(f"```{code_language}")
                else:
                    result.append("```")
            else:
                in_code_block = False
                result.append("```")
        elif in_code_block:
            result.append(line)
        else:
            result.append(line)

    return "\n".join(result)
